File: app/Models/Souscripteur.php
<?php

namespace App\Models;

use App\Models\Concerns\Auditable;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Illuminate\Support\Str;
use Laravel\Sanctum\HasApiTokens;

class Souscripteur extends Authenticatable
{
    use Auditable, HasApiTokens, Notifiable;

    public function auditLabel(): string
    {
        return trim($this->first_name . ' ' . $this->last_name) . ' (' . $this->uid . ')';
    }

    public const FRAIS_OUVERTURE_DEFAUT = 500000;

    protected $fillable = [
        'uid', 'first_name', 'last_name', 'email', 'phone', 'date_naissance',
        'id_type', 'id_number', 'address', 'photo',
        'password', 'app_access', 'last_login_at', 'statut',
        'frais_ouverture', 'frais_ouverture_payes', 'frais_ouverture_date',
    ];

    protected $hidden = [
        'password', 'remember_token',
    ];

    protected function casts(): array
    {
        return [
            'password' => 'hashed',
            'app_access' => 'boolean',
            'last_login_at' => 'datetime',
            'date_naissance' => 'date',
            'frais_ouverture' => 'decimal:2',
            'frais_ouverture_payes' => 'boolean',
            'frais_ouverture_date' => 'date',
        ];
    }

    public function souscriptions(): HasMany
    {
        return $this->hasMany(Souscription::class);
    }

    public function deviceTokens(): HasMany
    {
        return $this->hasMany(DeviceToken::class);
    }

    public function appNotifications(): HasMany
    {
        return $this->hasMany(ClientNotification::class)->latest();
    }

    public function fullName(): string
    {
        return trim($this->first_name . ' ' . $this->last_name);
    }

    public static function generateUid(): string
    {
        do {
            $uid = 'IMM-' . now()->year . '-' . strtoupper(Str::random(4));
        } while (self::where('uid', $uid)->exists());

        return $uid;
    }
}

File: mobile/lib/models/souscripteur.dart
class Souscripteur {
  final int id;
  final String uid;
  final String firstName;
  final String lastName;
  final String fullName;
  final String? email;
  final String? phone;
  final String? address;
  final String? photoUrl;
  final double fraisOuverture;
  final bool fraisOuverturePayes;
  final String? fraisRecuUrl;
  final int notificationsNonLues;

  Souscripteur({
    required this.id,
    required this.uid,
    required this.firstName,
    required this.lastName,
    required this.fullName,
    this.email,
    this.phone,
    this.address,
    this.photoUrl,
    this.fraisOuverture = 0,
    this.fraisOuverturePayes = false,
    this.fraisRecuUrl,
    this.notificationsNonLues = 0,
  });

  factory Souscripteur.fromJson(Map<String, dynamic> json) => Souscripteur(
        id: json['id'],
        uid: json['uid'] ?? '',
        firstName: json['first_name'] ?? '',
        lastName: json['last_name'] ?? '',
        fullName: json['full_name'] ?? '',
        email: json['email'],
        phone: json['phone'],
        address: json['address'],
        photoUrl: json['photo_url'],
        fraisOuverture: (json['frais_ouverture'] as num?)?.toDouble() ?? 0,
        fraisOuverturePayes: json['frais_ouverture_payes'] ?? false,
        fraisRecuUrl: json['frais_recu_url'],
        notificationsNonLues: (json['notifications_non_lues'] ?? 0) as int,
      );

  String get initials {
    final f = firstName.isNotEmpty ? firstName[0] : '';
    final l = lastName.isNotEmpty ? lastName[0] : '';
    return (f + l).toUpperCase();
  }
}
